### ISSUE TITLE
Current model is extremely slow to learn color information

### ISSUE THREAD

**Issue by ramayer** [ramayer/hierarchical-binary-quantization#11]
Chat history analyzing that issue is here: 

https://chatgpt.com/share/6a6fc15d-3870-83e8-8def-c9a67f93dfa3

> The model's now struggling to learn color.   In 24 hour of training, it was impressively good at structure and details (I guess thanks to LPIPS) but most of the images coming out are nearly monochrome (browns or greys).  In 12 more hours it didn't seem to improve much in either color or structure.  I had a similar problem before with a diffusion model where I can groupnorms in the wrong places; but I don't think it's the case this time.  Is it easy for you to check if color information should be able to flow through the model? I'm wondering if I can add a pixel-level color-emphasizing-loss-function and fine-tune from from where I am?
>
> ...
>
> LPIPS is very insensitive to exact color.
That's one of its strengths.
It mostly asks
"Does this look perceptually similar?"
not
"Did you reconstruct exactly the same shade of blue?"
>
> ...I think we should directly supervise saturation. 

```
def color_difference_loss(pred, target):
    return (
        F.l1_loss(pred[:,0]-pred[:,1], target[:,0]-target[:,1]) +
        F.l1_loss(pred[:,0]-pred[:,2], target[:,0]-target[:,2]) +
        F.l1_loss(pred[:,1]-pred[:,2], target[:,1]-target[:,2])
    )
```

**Comment by ramayer** [comment]
> ### User
>
> I think popular perceptual loss packages like lpips are quite color-insensitive.  They emphasize "an attractive girl's head" more than they do blond-vs-redhead (similar luminance); or hispanic-skin-tones vs asian-skin-tones (also similar luminance).
> 
> ### ChatGPT
> 
> I think that's a very plausible criticism of the standard perceptual losses.
> 
> LPIPS, VGG perceptual loss, DISTS, etc., were all developed around the idea of perceptual similarity rather than photometric accuracy. Human observers are much more tolerant of moderate hue shifts than they are of changes in structure, edges, identity, or texture. The metrics reflect that.
> 
> To take your examples:
> 
> A face with exactly the right geometry but slightly wrong hair color is still perceived as "the same face."
> A face with the right skin tone but the eyes shifted 4 pixels is immediately perceived as wrong.
> 
> So the learned feature extractors end up allocating much more representational power to geometry than to exact chromaticity.
> 
> For a generative model that's often exactly what you want.
> 
> For an autoencoder, though, the contract is different:
> 
> "Reproduce this image."
> 
> If the original dress is flame red, the reconstruction should be flame red.


The new Color loss term successfully for the other loss functions being mostly color insenstivite.
